src/storage/store/node_table.cpp

```diff
@@ -2,61 +2,97 @@
+            property, nodeGroupsDataFH, nodeGroupsMetaFH, &bufferManager, wal);
     }
-    return propertyColumns;
 }
 
-void NodeTable::initializeData(NodeTableSchema* nodeTableSchema) {
-    propertyColumns = initializeColumns(wal, &bufferManager, nodeTableSchema);
+void NodeTable::initializePKIndex(NodeTableSchema* nodeTableSchema) {
     if (nodeTableSchema->getPrimaryKey().dataType.getLogicalTypeID() != LogicalTypeID::SERIAL) {
         pkIndex = std::make_unique<PrimaryKeyIndex>(
             StorageUtils::getNodeIndexIDAndFName(wal->getDirectory(), tableID),
             nodeTableSchema->getPrimaryKey().dataType, bufferManager, wal);
     }
 }
 
-void NodeTable::scan(transaction::Transaction* transaction, ValueVector* inputIDVector,
-    const std::vector<uint32_t>& columnIds, std::vector<ValueVector*> outputVectors) {
-    assert(columnIds.size() == outputVectors.size());
+void NodeTable::read(transaction::Transaction* transaction, ValueVector* inputIDVector,
+    const std::vector<column_id_t>& columnIds, const std::vector<ValueVector*>& outputVectors) {
+    if (inputIDVector->isSequential()) {
+        scan(transaction, inputIDVector, columnIds, outputVectors);
+    } else {
+        lookup(transaction, inputIDVector, columnIds, outputVectors);
+    }
+}
+
+void NodeTable::write(common::property_id_t propertyID, common::ValueVector* nodeIDVector,
+    common::ValueVector* vectorToWriteFrom) {
+    assert(propertyColumns.contains(propertyID));
+    propertyColumns.at(propertyID)->write(nodeIDVector, vectorToWriteFrom);
+}
+
+void NodeTable::scan(Transaction* transaction, ValueVector* inputIDVector,
+    const std::vector<column_id_t>& columnIds, const std::vector<ValueVector*>& outputVectors) {
+    assert(columnIds.size() == outputVectors.size() && !inputIDVector->state->isFlat());
     for (auto i = 0u; i < columnIds.size(); i++) {
-        if (columnIds[i] == UINT32_MAX) {
+        if (columnIds[i] == INVALID_COLUMN_ID) {
             outputVectors[i]->setAllNull();
         } else {
-            propertyColumns.at(columnIds[i])->read(transaction, inputIDVector, outputVectors[i]);
+            propertyColumns.at(columnIds[i])->scan(transaction, inputIDVector, outputVectors[i]);
         }
     }
 }
 
-offset_t NodeTable::addNodeAndResetProperties() {
-    auto nodeOffset = nodesStatisticsAndDeletedIDs->addNode(tableID);
-    for (auto& [_, column] : propertyColumns) {
-        if (column->dataType.getLogicalTypeID() != LogicalTypeID::SERIAL) {
-            column->setNull(nodeOffset);
+void NodeTable::lookup(Transaction* transaction, ValueVector* inputIDVector,
+    const std::vector<column_id_t>& columnIds, const std::vector<ValueVector*>& outputVectors) {
+    assert(columnIds.size() == outputVectors.size());
+    auto pos = inputIDVector->state->selVector->selectedPositions[0];
+    for (auto i = 0u; i < columnIds.size(); i++) {
+        if (columnIds[i] == INVALID_COLUMN_ID) {
```

Comment:
This is triggered by `ScanMultiNodeTables` when a column doesn't exist in a table. Maybe this logic should be moved inside `ScanMultiNodeTables`.

Reply:
Can you open an issue about this, so it's at least in our records. We'll otherwise forget about it. Or do the necessary change so we get done with it.

Reply:
Opened it.